docs/DafnyRef/Types.md

```diff
@@ -1705,29 +1776,32 @@ such a `C` object can be used as a value of type `J`.
 
 ## 13.1. Type `object`
 ````grammar
-ObjectType_ = "object"
+ObjectType_ = "object" | "object?"
 ````
 
-There is a built-in trait `object` that is a supertype of all
-reference types.
+There is a built-in trait `object?` that is a supertype of all
+reference types and a trait `object` that is a supertype of all non-null reference types..
 Every class and every trait (other than `object` itself) automatically extends
-`object`. This includes types like arrays and iterators that do not permit
-explicit extending of traits. The purpose of type `object`
+`object?`; every non-null class or trait extends `object`.
+This includes types like arrays and iterators that do not permit
+explicit extending of traits. The purpose of type `object
 is to enable a uniform treatment of _dynamic frames_. In particular, it
 is useful to keep a ghost field (typically named `Repr` for
 "representation") of type `set<object>`.
 
-It serves no purpose (but does no harm) to explicitly list `object` as
+It serves no purpose (but does no harm) to explicitly list `object` ior `object?` as
```

Comment:
No, `object?` is not allowed to be listed in any `extends` clause. What goes into the `extends` clause is the name of the `trait` itself (suitably parameterized), not the _trait type_. In this, I'm saying "`trait` type" to mean the name of the type without the `?` at the end, and "trait type" to mean the name with a `?` at the end.

Reply:
OK